gh-117511: Make PyMutex public in the non-limited API

colesbury opened the pull request:
<!-- gh-issue-number: gh-117511 -->
* Issue: gh-117511
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Direct link 📚: https://cpython-previews--117731.org.readthedocs.build/en/117731/c-api/init.html#synchronization-primitives
📚 Documentation preview 📚: https://cpython-previews--117731.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

colesbury committed:
gh-117511: Make PyMutex public in the non-limited API

colesbury committed:
Add blurb

colesbury committed:
Add missing newline

colesbury committed:
Replace `struct _PyMutex` with `PyMutex`

erlend-aasland commented on Doc/c-api/init.rst:
```diff
@@ -1942,3 +1947,31 @@ be used in new code.
+
+.. c:type:: PyMutex
+
+   A mutual exclusion lock.  The ``PyMutex`` should be initialized to zero to
```
```suggestion
   A mutual exclusion lock.  The :c:type:`!PyMutex` should be initialized to zero to
```

erlend-aasland commented on Doc/c-api/init.rst:
```diff
@@ -1942,3 +1947,31 @@ be used in new code.
+
+.. c:function:: void PyMutex_Lock(PyMutex *m)
+
+   Lock the mutex.  If another thread has already locked it, the calling
```
```suggestion
   Lock mutex *m*.  If another thread has already locked it, the calling
```

erlend-aasland commented on Doc/c-api/init.rst:
```diff
@@ -1942,3 +1947,31 @@ be used in new code.
+
+.. c:function:: void PyMutex_Unlock(PyMutex *m)
+
+   Unlock the mutex. The mutex must be locked --- otherwise, the function will
```
```suggestion
   Unlock mutex *m*. The mutex must be locked --- otherwise, the function will
```

erlend-aasland commented on Misc/NEWS.d/next/C API/2024-04-10-16-48-04.gh-issue-117511.RZtBRK.rst:
```diff
@@ -0,0 +1 @@
+Make ``PyMutex`` public in the non-limited C API.
```
```suggestion
Make the :c:type:`PyMutex` public in the non-limited C API.
```

colesbury committed:
Apply suggestions from code review

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

colesbury committed:
Update Doc/c-api/init.rst

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

colesbury committed:
Use underscore prefix for PyMutex field.

Update code comments to further clarify that the `_bits` field is not
part of the public C API.

colesbury committed:
Fix comment

colesbury committed:
PyMutex should not be copied or moved

colesbury committed:
Merge branch 'main' into gh-117511-pymutex

colesbury committed:
Update documentation

colesbury committed:
Export PyMutex_Lock and PyMutex_Unlock as regular functions

vstinner commented on Include/cpython/lock.h:
```diff
@@ -30,32 +30,34 @@ typedef struct PyMutex {
     uint8_t expected = _Py_UNLOCKED;
     if (!_Py_atomic_compare_exchange_uint8(&m->_bits, &expected, _Py_LOCKED)) {
-        _PyMutex_LockSlow(m);
+        PyMutex_Lock(m);
```
Why not keeping _PyMutex_LockSlow()? Is it a recursive call? It's unclear to me.

I would expect PyMutex_Lock() to do the exactly the same than the static inline function.

colesbury replied:
It doesn't matter much, but it's convenient to make the "slow path" and the non-inline function the same function.

The inline function (`_PyMutex_Lock`) is written to generate few instructions because it's inlined in lots of places, but that consideration doesn't make as much sense for the non-inline function.

colesbury replied:
To be more clear about this, the non-inlined function starts with a more general handling of `unlocked -> locked` transition that also handles the case where there's a waiter (but the lock is unlocked):

https://github.com/python/cpython/blob/35b16795d11cb50768ffad5fe8e61bdebde9b66a/Python/lock.c#L59-L64

The inline function only handles the specific case where the lock is unlocked AND there are no waiters:

https://github.com/python/cpython/blob/35b16795d11cb50768ffad5fe8e61bdebde9b66a/Include/internal/pycore_lock.h#L73-L74

The more general case is ~3 extra instructions. It's usually worth avoiding that in the inline path, but that consideration doesn't really matter in the non-inline case.

vstinner replied:
Ok I see. It's surprising, but it makes sense. Also, it's up to you :-)

vstinner commented on Include/lock.h:
```diff
@@ -0,0 +1,16 @@
+
+#ifndef Py_LIMITED_API
+#  define Py_CPYTHON_LOCK_H
+#  include "cpython/lock.h"
```
I don't see the value of such header file. Just include cpython/lock.h in Python.h, and check Py_LIMITED_API in cpython/lock.h.

colesbury replied:
@ericsnowcurrently expressed a preference for this style when `pyatomic.h` was added: https://github.com/python/cpython/pull/109344#discussion_r1325076002

ericsnowcurrently replied:
IMO, it's helpful to follow a consistent pattern when it comes to the Include/cpython header files.  That means in some cases we end up with very minimal header files like this in Include/.

vstinner replied:
If you mention consistency, there are already many header files in Include/cpython/ which have no companion Include/ header file:

```
$ grep cpython/ Include/Python.h
#include "cpython/pydebug.h"
#include "cpython/longintrepr.h"
#include "cpython/odictobject.h"
#include "cpython/funcobject.h"
#include "cpython/classobject.h"
#include "cpython/code.h"
#include "cpython/cellobject.h"
#include "cpython/initconfig.h"
#include "cpython/genobject.h"
#include "cpython/picklebufobject.h"
#include "cpython/pytime.h"
#include "cpython/context.h"
#include "cpython/pyctype.h"
#include "cpython/pyfpe.h"
#include "cpython/tracemalloc.h"
#include "cpython/optimizer.h"
```

vstinner commented on Objects/object.c:
```diff
@@ -2372,7 +2372,7 @@ new_reference(PyObject *op)
     op->ob_tid = _Py_ThreadId();
     op->_padding = 0;
-    op->ob_mutex = (struct _PyMutex){ 0 };
+    op->ob_mutex = (PyMutex){ 0 };
```
If there is no public PyMutex_STATIC_INIT, can you maybe add a private one in pycore_lock.h?

colesbury replied:
Let's consider that in a separate PR:

* It's not directly related to making the API public
* It would make sense to consider that for multiple types (e.g., `PyEvent`, `_PyOnceFlag`), not just `PyMutex`

vstinner commented on Python/lock.c:
```diff
@@ -47,18 +47,19 @@ _Py_yield(void)
+#undef PyMutex_Lock
 void
-_PyMutex_LockSlow(PyMutex *m)
+PyMutex_Lock(PyMutex *m)
```
You might move it to the end to still use the static inline macro in the following lines.

vstinner commented on Doc/c-api/init.rst:
```diff
@@ -2152,3 +2157,41 @@ be used in new code.
+      considered unstable.  The size may change in future Python releases
+      without a deprecation period.
+
+   .. versionadded:: 3.13
```
We are past 3.13 beta1. Did you get an exception to get this feature in Python 3.13? If no, you should target Python 3.14.

colesbury replied:
@Yhg1s is it acceptable to expose `PyMutex` as a public C API in 3.13? There are docs and refactoring to make it public, but no new functionality.

(For context, the only other free-threading related API I'd like to see public in 3.13 is the [critical section API](https://github.com/python/cpython/issues/119344), but there isn't a decision yet on that from the WG)

Yhg1s replied:
Yes, considering the use-case, the extent of the discussion around it, and the fact that this is exposing an existing (albeit new in 3.13) API, this is fine to go in before beta 3.

colesbury committed:
Move PyMutex_Lock/Unlock definitions to end of file

vstinner commented on Doc/c-api/init.rst:
```diff
@@ -55,6 +55,11 @@ The following functions can be safely called before Python is initialized:
   * :c:func:`PyMem_RawCalloc`
   * :c:func:`PyMem_RawFree`
 
+* Synchronization
```
```suggestion
* Synchronization:
```

vstinner committed:
Update Doc/c-api/init.rst

vstinner commented on Doc/c-api/init.rst:
```diff
@@ -2152,3 +2157,41 @@ be used in new code.
+.. c:function:: void PyMutex_Unlock(PyMutex *m)
+
+   Unlock mutex *m*. The mutex must be locked --- otherwise, the function will
+   issue a fatal error.
```
I would prefer to emit an unraisable exception, log a warning, or anything else. A fatal error kills the process, it's not great when Python is embedded. But not ignoring bugs silently is a good thing :-)

vstinner replied:
I change my mind, it's not possible, see the comment below on lock.c.

vstinner commented on Python/lock.c:
```diff
@@ -584,3 +570,19 @@ uint32_t _PySeqLock_AfterFork(_PySeqLock *seqlock)
+PyMutex_Unlock(PyMutex *m)
+{
+    if (_PyMutex_TryUnlock(m) < 0) {
+        Py_FatalError("unlocking mutex that is not locked");
```
I suggest to only issue a fatal error if Python is built in debug mode or in Python Development Mode (-X dev). Otherwise, emit a warning. If the warning raises an exception (-Werror), log the warning as an unraisable exception.

```suggestion
        int fatal_error;
#ifdef Py_DEBUG
        fatal_error = 1;
#else
        fatal_error = _Py_GetConfig()->dev_mode;
#endif
        if (fatal_error) {
            Py_FatalError("unlocking mutex that is not locked");
        }
        else {
            if (PyErr_WarnEx(PyExc_SystemError,
                    "unlocking mutex that is not locked", 1))
            {
                PyErr_WriteUnraisable(NULL);
            }
        }
```

zooba replied:
Mutexes are too low level to handle this way. We voted against requiring a return value to indicate failure, which means the only viable option is to tear down the entire process with a fatal error.

colesbury replied:
Yeah, it's not safe to issue a warning here because the you might not have a valid thread state / hold the GIL. I don't think it's worth trying to do something more clever than `Py_FatalError` here because you run the risk of your error reporting mechanism crashing and hiding the original programmer error.

vstinner replied:
Oh ok, that's sad, but I get it :-)

vstinner reviewed (approved):
LGTM. Thanks for the different update and the doc.

@colesbury got an exception from @Yhg1s (3.13 release manager) to add this API to Python 3.13, god.
